### PR TITLE
User-1: add initial user functionality

### PR DESCRIPTION
**Pull request by Sunagatov**
Changes:
1.`user_details` table was created with liquabase, jpaReposity and entity class.
2. sql inserts for `user_details`  table were added using liquabase.
3. `address` table was created with liquabase, jpaReposity and entity class.
4. sql inserts for `address`  table were added using liquabase.
5. getUserById endpoint was created.
6. JWT registration and auth were updated with the new user services. Now after registration user_details are saved in database, before they were stored in the memory.

Tests:
1. I registered a new user with the JWT registration endpoint using Postman. New JWT token was returned as expected. Then I checked that this user was saved in the database as expected.
2. I authenticated the existed user using his username and password with the JWT auth endpoint using Postman. New JWT token was returned as expected. 
3. I used the JWT from the previous steps to test the `getUserById` endpoint using Postman. `UserDetails` json was returned as expected.

Conclusion:
Project was built with Apache Maven successfully.
Project was tested manually with Postman successfully.

### CODE REVIEW

File: src/main/java/com/zufar/onlinestore/security/authentication/UserDetailsServiceImpl.java [hunk 1]
```diff
@@ -1,23 +1,46 @@
...
+    @Override
+    public UserDetails loadUserByUsername(final String username) throws UsernameNotFoundException {
+        UserEntity user = userRepository.findUserByUserName(username);
+        if (user == null) {
+            log.warn("Failed to get the user with the userName = {}.", username);
+            throw new UserNotFoundException(username);
+        }
+        UserDto userDto = userDtoConverter.toDto(user);
+        SimpleGrantedAuthority userAuthority = new SimpleGrantedAuthority("User");
+        final Set<GrantedAuthority> authorities = Set.of(userAuthority);
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/security/authentication/UserDetailsServiceImpl.java:39]: maybe use `Collections.singleton()` to explicitly define that this set will contain only one element

File: src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.util.Set;
+
+@Service
+@RequiredArgsConstructor
+public class RegistrationDtoConverter {
+
+    private final PasswordEncoder passwordEncoder;
+
+
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java:20]: extra space

File: src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                registrationRequest.lastName(),
+                registrationRequest.userName(),
+                registrationRequest.email(),
+                encodedPassword,
+                registrationRequest.address()
+        );
+    }
+
+    public User toUser(RegistrationRequest registrationRequest) {
+        SimpleGrantedAuthority userAuthority = new SimpleGrantedAuthority("User");
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java:36]: don't use plain text

File: src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        final Set<GrantedAuthority> authorities = Set.of(userAuthority);
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java:37]: same here

File: src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        String userName,
+
+        @Email(message = "Email should be valid")
+        @NotBlank(message = "Email is the mandatory attribute")
+        String email,
+
+        @NotBlank(message = "Email is the mandatory attribute")
+        String password,
+
+        @Valid
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java:32]: i think it doesn't make a sense to put `@Valid` here

File: src/main/java/com/zufar/onlinestore/user/api/UserService.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+    @Override
+    public UserDto saveUser(final UserDto saveUserRequest) {
+        UserEntity userEntity = userDtoConverter.toEntity(saveUserRequest);
+        UserEntity userEntityWithId = userCrudRepository.save(userEntity);
+        return userDtoConverter.toDto(userEntityWithId);
+    }
+
+    @Override
+    public UserDto getUserById(final UUID userId) throws UserNotFoundException {
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/api/UserService.java:31]: @ Override
public UserDto getUserById(final UUID userId) throws UserNotFoundException {
    return userCrudRepository.findById(userId)
        .map(userDtoConverter::toDto)
        .orElseThrow(() -> {
            log.warn("Failed to get the user with the userId = {}.", userId);
            return new UserNotFoundException(userId);
        });
}

File: src/main/java/com/zufar/onlinestore/user/converter/AddressDtoConverter.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package com.zufar.onlinestore.user.converter;
+
+import com.zufar.onlinestore.user.dto.AddressDto;
+import com.zufar.onlinestore.user.entity.Address;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/converter/AddressDtoConverter.java:7]: Use `@Component` here. Also I suppose that we will create a separate factory for more efficient bean management in the future

File: src/main/java/com/zufar/onlinestore/user/converter/UserDtoConverter.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+public class UserDtoConverter {
+    private final AddressDtoConverter addressDtoConverter;
+
+    public UserDto toDto(final UserEntity entity) {
+        AddressDto addressDto = addressDtoConverter.toDto(entity.getAddress());
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/converter/UserDtoConverter.java:16]: can entity address be null?

File: src/main/java/com/zufar/onlinestore/user/converter/UserDtoConverter.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+                entity.getLastName(),
+                entity.getUserName(),
+                entity.getEmail(),
+                entity.getPassword(),
+                addressDto
+        );
+    }
+
+    public UserEntity toEntity(final UserDto dto) {
+        Address address = addressDtoConverter.toEntity(dto.address());
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/converter/UserDtoConverter.java:29]: same here

File: src/main/java/com/zufar/onlinestore/user/dto/UserDto.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+        @Email(message = "Email should be valid")
+        @NotBlank(message = "Email is the mandatory attribute")
+        String email,
+
+        @Email(message = "Password should be valid")
+        @NotBlank(message = "Password is the mandatory attribute")
+        String password,
+
+        @Valid
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/dto/UserDto.java:39]: i think it does not necessarily to put Valid here

File: src/main/java/com/zufar/onlinestore/user/endpoint/UserEndpoint.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.util.UUID;
+
+
+@Slf4j
+@RestController
+@RequiredArgsConstructor
+@Validated
+@RequestMapping(value = UserEndpoint.API_CUSTOMERS)
+public class UserEndpoint {
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/endpoint/UserEndpoint.java:22]: change name for more specific. It is a bad practice to name class that expose and implement API as 'Endpoint'

File: src/main/java/com/zufar/onlinestore/user/endpoint/UserEndpoint.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+    public static final String API_CUSTOMERS = "/api/v1/users";
+
+    private final UserApi userApi;
+
+    @GetMapping("/{userId}")
+    public ResponseEntity<UserDto> getUserById(@PathVariable("userId") final String userId) {
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/endpoint/UserEndpoint.java:29]: you can leave "userId" in PathVariable just with `@PathVariable`

File: src/main/java/com/zufar/onlinestore/user/entity/UserEntity.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+import java.util.UUID;
+
+
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/entity/UserEntity.java:21]: extra space

File: src/main/java/com/zufar/onlinestore/user/exception/UserNotFoundException.java [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.UUID;
+
+@Getter
+public class UserNotFoundException extends RuntimeException {
+
+    private final UUID userId;
+    private final String userName;
+
+    public UserNotFoundException(UUID userId) {
+        super(String.format("User with id = %s is not found.", userId));
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/user/exception/UserNotFoundException.java:14]: maybe just leave only one version of constructor with id?

File: src/main/resources/db/changelog/version-1.0/28.07.2023.part1.create-address-table.sql [hunk 15]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE address
+(
+    id      UUID         NOT NULL PRIMARY KEY,
+    line    VARCHAR(255) NOT NULL,
+    city    VARCHAR(255) NOT NULL,
```
Comment (yevr19) [inline src/main/resources/db/changelog/version-1.0/28.07.2023.part1.create-address-table.sql:5]: please check standarts for defining length on fields. It's too much right now

File: src/main/resources/db/changelog/version-1.0/28.07.2023.part3.create-user-details-table.sql [hunk 16]
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE user_details
+(
+    id         UUID         NOT NULL PRIMARY KEY,
+    first_name VARCHAR(255) NOT NULL,
+    last_name  VARCHAR(255) NOT NULL,
+    user_name  VARCHAR(255) NOT NULL,
```
Comment (yevr19) [inline src/main/resources/db/changelog/version-1.0/28.07.2023.part3.create-user-details-table.sql:6]: the same here, check length

Comment (Sunagatov) [reply src/main/resources/db/changelog/version-1.0/28.07.2023.part3.create-user-details-table.sql:6]: fixed

File: src/main/resources/db/changelog/version-1.0/28.07.2023.part1.create-address-table.sql (hunk 15 above)
Comment (Sunagatov) [reply src/main/resources/db/changelog/version-1.0/28.07.2023.part1.create-address-table.sql:5]: fixed

File: src/main/java/com/zufar/onlinestore/user/exception/UserNotFoundException.java (hunk 14 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/user/exception/UserNotFoundException.java:14]: fixed

File: src/main/java/com/zufar/onlinestore/user/endpoint/UserEndpoint.java (hunk 12 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/user/endpoint/UserEndpoint.java:29]: fixed

File: src/main/java/com/zufar/onlinestore/security/authentication/UserDetailsServiceImpl.java [hunk 17]
```diff
@@ -1,23 +1,46 @@
...
+
+    @Override
+    public UserDetails loadUserByUsername(final String username) throws UsernameNotFoundException {
+        UserEntity user = userRepository.findUserByUserName(username);
+        if (user == null) {
+            log.warn("Failed to get the user with the userName = {}.", username);
+            throw new UserNotFoundException(username);
+        }
+        UserDto userDto = userDtoConverter.toDto(user);
+        SimpleGrantedAuthority userAuthority = new SimpleGrantedAuthority("User");
```
[the remark this answers is not in the document]
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/security/authentication/UserDetailsServiceImpl.java:37]: Fixed

File: src/main/java/com/zufar/onlinestore/security/authentication/UserDetailsServiceImpl.java (hunk 1 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/security/authentication/UserDetailsServiceImpl.java]: Fixed

File: src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java (hunk 3 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java:33]: Fixed

File: src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java (hunk 2 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java]: Fixed

File: src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java (hunk 4 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/security/converter/RegistrationDtoConverter.java:34]: Fixed

File: src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        String username,
+
+        @Email(message = "Email should be valid")
+        @NotBlank(message = "Email is the mandatory attribute")
+        String email,
+
+        @NotBlank(message = "Email is the mandatory attribute")
+        String password,
+
+        @Valid
```
Comment (AiratGimaev) [inline src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java:32]: I think it makes sense)
But better check it)

File: src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java (hunk 5 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java:32]: QA team will double check it

File: src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java (hunk 18 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/security/dto/authentication/RegistrationRequest.java:32]: QA team will double check it

**Review by yevr19 — APPROVED** [review]
approved

**Pull request merged**